packages/flutter/lib/src/rendering/object.dart

```diff
@@ -651,31 +651,58 @@ class PaintingContext extends ClipContext {
     }
   }
 
-  /// Blend further painting with an alpha value.
+  /// Blend further painting with an opacity value.
   ///
-  /// The `offset` argument indicates an offset to apply to all the children
-  /// (the rendering created by `painter`).
+  /// {@macro flutter.rendering.PaintingContext.pushClipRect.needsCompositing}
+  ///
+  /// The [offset] argument indicates an offset to apply to all the children
+  /// (the rendering created by [painter]).
+  ///
+  /// The [size] is used to compute the painting bounds if a save layer is used.
+  /// If the [size] is not provided, no estimate is used.
   ///
-  /// The `alpha` argument is the alpha value to use when blending the painting
-  /// done by `painter`. An alpha value of 0 means the painting is fully
-  /// transparent and an alpha value of 255 means the painting is fully opaque.
+  /// The [opacity] argument is the opacity value to use when blending the painting
+  /// done by [painter]. An opacity value of 0.0 means the painting is fully
+  /// transparent and an opacity value of 1.0 means the painting is fully opaque.
   ///
-  /// The `painter` callback will be called while the `alpha` is applied. It
+  /// The [painter] callback will be called while the [opacity] is applied. It
   /// is called synchronously during the call to [pushOpacity].
   ///
-  /// {@macro flutter.rendering.PaintingContext.pushClipRect.oldLayer}
+  /// This method will only create a layer if [needsCompositing] is `true` and
+  /// [opacity] is greater than 0 and less than 1.0.
   ///
-  /// A [RenderObject] that uses this function is very likely to require its
-  /// [RenderObject.alwaysNeedsCompositing] property to return true. That informs
-  /// ancestor render objects that this render object will include a composited
-  /// layer, which, for example, causes them to use composited clips.
-  OpacityLayer pushOpacity(Offset offset, int alpha, PaintingContextCallback painter, { OpacityLayer? oldLayer }) {
-    final OpacityLayer layer = oldLayer ?? OpacityLayer();
-    layer
-      ..alpha = alpha
-      ..offset = offset;
-    pushLayer(layer, painter, Offset.zero);
-    return layer;
+  /// {@macro flutter.rendering.PaintingContext.pushClipRect.oldLayer}
+  OpacityLayer? pushOpacity(bool needsCompositing, Offset offset, double opacity, PaintingContextCallback painter, { OpacityLayer? oldLayer, Size? size }) {
+    assert(opacity >= 0 && opacity <= 1);
+    if (opacity <= 0) {
+      return null;
+    }
+    if (opacity >= 1.0) {
+      painter(this, offset);
+      return null;
+    }
+    // Due to https://github.com/flutter/flutter/issues/48417 this will always need to be
+    // composited on the web.
```

Comment:
can't we just fix https://github.com/flutter/flutter/issues/48417 first?

Reply:
if we could easily detect canvaskit mode, this would only be an issue on HTML renderer. @yjbanov for the feasibility of either fixing that or giving us a better flag

Reply:
from discussion offline, this isn't fixable in html mode.

Reply:
(not fixable with the current renderer)